File: Source/AlphaTab/Rendering/Glyphs/TabTimeSignatureGlyph.cs
namespace AlphaTab.Rendering.Glyphs
{
    internal class TabTimeSignatureGlyph : TimeSignatureGlyph
    {
        protected override float CommonY
        {
            get
            {
                var renderer = (TabBarRenderer)Renderer;
                return renderer.GetTabY(0);
            }
        }

        protected override float NumeratorY
        {
            get
            {
                var renderer = (TabBarRenderer)Renderer;
                var offset = renderer.Bar.Staff.Tuning.Length <= 4 ? 1 / 4f : 1 / 3f;
                return renderer.LineOffset * renderer.Bar.Staff.Tuning.Length * offset * Scale;
            }
        }

        protected override float DenominatorY
        {
            get
            {
                var renderer = (TabBarRenderer)Renderer;
                const float offset = 3 / 5f;
                return renderer.LineOffset * renderer.Bar.Staff.Tuning.Length * offset * Scale;
            }
        }

        protected override float CommonScale => 1;

        protected override float NumberScale
        {
            get
            {
                var renderer = (TabBarRenderer)Renderer;
                if (renderer.Bar.Staff.Tuning.Length <= 4)
                {
                    return 0.75f;
                }

                return 1;
            }
        }

        public TabTimeSignatureGlyph(float x, float y, int numerator, int denominator, bool isCommon)
            : base(x, y, numerator, denominator, isCommon)
        {
        }
    }
}

File: packages/alphatab/src/rendering/glyphs/TabTimeSignatureGlyph.ts
import { EngravingSettings } from '@coderline/alphatab/EngravingSettings';
import { BarSubElement } from '@coderline/alphatab/model/Bar';
import { TimeSignatureGlyph } from '@coderline/alphatab/rendering/glyphs/TimeSignatureGlyph';
import type { TabBarRenderer } from '@coderline/alphatab/rendering/TabBarRenderer';

/**
 * @internal
 */
export class TabTimeSignatureGlyph extends TimeSignatureGlyph {
    public override doLayout(): void {
        this.barSubElement = BarSubElement.GuitarTabsTimeSignature;
        super.doLayout();
    }

    protected get commonScale(): number {
        return 1;
    }

    protected get numberScale(): number {
        const renderer: TabBarRenderer = this.renderer as TabBarRenderer;
        if (renderer.bar.staff.tuning.length <= 4) {
            return EngravingSettings.GraceScale;
        }
        return 1;
    }
}
